firebase-vscode/webviews/EmulatorPanel.tsx

```diff
@@ -0,0 +1,249 @@
+  );
+  function setEmulatorUiSelectionsAndSaveToWorkspace(
+    uiSelections: EmulatorUiSelections
+  ) {
+    // TODO(christhompson): Save UI selections in the current workspace.
+    // Requires context object.
+    setEmulatorUiSelections(uiSelections);
+  }
+  const [showEmulatorProgressIndicator, setShowEmulatorProgressIndicator] =
+    useState<boolean>(false);
+
+  // TODO(christhompson): Load UI selections from the current workspace.
+  // Requires context object.
+  // TODO(christhompson): Check if the emulators are running on extension start.
+  const [runningEmulatorInfo, setRunningEmulatorInfo] =
+    useState<RunningEmulatorInfo>();
+
+  broker.on("notifyEmulatorsStopped", () => {
+    setShowEmulatorProgressIndicator(false);
+    console.log(`notifyEmulatorsStopped received in sidebar`);
+    setRunningEmulatorInfo(null);
+  });
+
+  broker.on("notifyRunningEmulatorInfo", (info: RunningEmulatorInfo) => {
+    setShowEmulatorProgressIndicator(false);
+    console.log(`notifyRunningEmulatorInfo received in sidebar`);
+    setRunningEmulatorInfo(info);
+  });
+
+  broker.on("notifyEmulatorImportFolder", ({ folder }) => {
+    console.log(`notifyEmulatorImportFolder received in sidebar: ${folder}`);
+    emulatorUiSelections.importStateFolderPath = folder;
+    setEmulatorUiSelectionsAndSaveToWorkspace({ ...emulatorUiSelections }); // rerender clone
+  });
+
+  function launchEmulators() {
+    if (!emulatorUiSelections.projectId) {
+      broker.send("showMessage", {
+        msg: "Missing project ID",
+        options: {
+          modal: true,
+          detail: `Please specify a project ID before starting the emulator suite.`,
+        },
+      });
+      return;
+    }
+    if (!firebaseJson) {
+      // TODO(christhompson): Consider using a default config in the case that
+      // firebase.json doesnt exist.
+      broker.send("showMessage", {
+        msg: "Missing firebase.json",
+        options: {
+          modal: true,
+          detail: `Unable to find firebase.json file.`,
+        },
+      });
+      return;
+    }
+    setShowEmulatorProgressIndicator(true);
+    broker.send("launchEmulators", {
+      firebaseJson,
+      emulatorUiSelections,
+    });
+  }
+
+  function stopEmulators() {
+    setShowEmulatorProgressIndicator(true);
+    broker.send("stopEmulators");
+  }
+
+  /**
+   * Called when import folder changes.
+   */
+  function selectedImportFolder(event: any) {
+    event.preventDefault();
+    broker.send("selectEmulatorImportFolder");
+  }
+
+  function toggleExportOnExit() {
+    console.log("toggle export on exit");
+    const selections: EmulatorUiSelections = emulatorUiSelections;
+    selections.exportStateOnExit = !selections.exportStateOnExit;
+    setEmulatorUiSelectionsAndSaveToWorkspace(selections);
+  }
+
+  function projectIdChanged(event: React.ChangeEvent<HTMLInputElement>) {
+    console.log("projectIdChanged: " + event.target.value);
```

Comment:
Use webLogger for these? FYI it doesn't do `.log` because the treatment of `.log()` in the CLI logger seemed a little different and I didn't know what was up with it, so `info` or `debug` maybe.

Reply:
Whoops, thanks for pointing that out, these should have been removed/logged already